qutip/random_objects.py

```diff
@@ -212,7 +214,7 @@ def _rand_herm_dense(N, density, pos_def):
         M[row, col] = 0
     if pos_def:
         as_vec = M.ravel()
-        M[::N+1] = (np.abs(M.diagonal())+np.sqrt(2)*N)
+        as_vec[::N+1] = (np.abs(M.diagonal()) + np.sqrt(2)*N)
```

Comment:
This looks unsafe without a reassignment back to `M`:
```python
>>> x = np.array(np.eye(2), order='F')
>>> np.shares_memory(x.ravel(), x)
False
```

`np.ravel` doesn't guarantee that it returns a view.  We'd have the same C/Fortran problem with `np.ndarray.flat` and `np.reshape` as well, so we probably just have to reassign back to `M` to be safe.  I imagine we'll never see problems in reality because Numpy always defaults to C-order, but it's probably best not to rely on it just in case.  At any rate, reassignment to `M` is a free operation.

(*edit*: actually, on second thoughts, we can probably just use `np.fill_diagonal`)